src/ConnectedMode/Binding/IUnintrusiveBindingController.cs

```diff
@@ -29,32 +30,57 @@ namespace SonarLint.VisualStudio.ConnectedMode.Binding
     internal interface IUnintrusiveBindingController
     {
         Task BindAsync(BoundSonarQubeProject project, IProgress<FixedStepsProgress> progress, CancellationToken token);
+        Task BindAsync(BoundServerProject project, IProgress<FixedStepsProgress> progress, CancellationToken token);
     }
 
     [Export(typeof(IUnintrusiveBindingController))]
     [PartCreationPolicy(CreationPolicy.NonShared)]
     internal class UnintrusiveBindingController : IUnintrusiveBindingController
     {
         private readonly IBindingProcessFactory bindingProcessFactory;
+        private readonly IServerConnectionsRepository serverConnectionsRepository;
+        private readonly ISolutionInfoProvider solutionInfoProvider;
 
         [ImportingConstructor]
-        public UnintrusiveBindingController(IBindingProcessFactory bindingProcessFactory)
+        public UnintrusiveBindingController(IBindingProcessFactory bindingProcessFactory, IServerConnectionsRepository serverConnectionsRepository, ISolutionInfoProvider solutionInfoProvider)
         {
             this.bindingProcessFactory = bindingProcessFactory;
+            this.serverConnectionsRepository = serverConnectionsRepository;
+            this.solutionInfoProvider = solutionInfoProvider;
         }
 
-        public async Task BindAsync(BoundSonarQubeProject project, IProgress<FixedStepsProgress> progress, CancellationToken token)
+        public async Task BindAsync(BoundServerProject project, IProgress<FixedStepsProgress> progress, CancellationToken token)
         {
             var bindingProcess = CreateBindingProcess(project);
-
             await bindingProcess.DownloadQualityProfileAsync(progress, token);
             await bindingProcess.SaveServerExclusionsAsync(token);
         }
 
-        private IBindingProcess CreateBindingProcess(BoundSonarQubeProject project)
+        public async Task BindAsync(BoundSonarQubeProject project, IProgress<FixedStepsProgress> progress, CancellationToken token)
+        {
+            var proposedConnection = ServerConnection.FromBoundSonarQubeProject(project);
+
+            if (proposedConnection is null)
+            {
+                throw new InvalidOperationException(BindingStrings.UnintrusiveController_InvalidConnection);
+            }
+            
+            if (!serverConnectionsRepository.TryGet(proposedConnection.Id, out var connection))
+            {
+                if (!serverConnectionsRepository.TryAdd(proposedConnection))
```

Comment:
I think we should document here why we are adding the connection, in case it doesn't exist.

Reply:
We add it because it doesn't exist and it's a migration process🙃

Reply:
Exactly that you should document.

Reply:
I can rename the method to mention migration specifically.